src/relations/db.py

```diff
@@ -0,0 +1,258 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Postgres db and db-admin relation hooks & helpers."""
+
+
+import logging
+from typing import Iterable
+
+from charms.postgresql_k8s.v0.postgresql import (
+    PostgreSQLCreateDatabaseError,
+    PostgreSQLCreateUserError,
+    PostgreSQLDeleteUserError,
+    PostgreSQLGetPostgreSQLVersionError,
+)
+from ops.charm import (
+    CharmBase,
+    RelationBrokenEvent,
+    RelationChangedEvent,
+    RelationDepartedEvent,
+)
+from ops.framework import Object
+from ops.model import BlockedStatus, Relation, Unit
+from pgconnstr import ConnectionString
+
+from utils import new_password
+
+logger = logging.getLogger(__name__)
+
+
+class DbProvides(Object):
+    """Defines functionality for the 'provides' side of the 'db' relation.
+
+    Hook events observed:
+        - relation-changed
+        - relation-departed
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase, admin: bool = False):
+        """Constructor for DbProvides object.
+
+        Args:
+            charm: the charm for which this relation is provided
+            admin: a boolean defining whether or not this relation has admin permissions, switching
+                between "db" and "db-admin" relations.
+        """
+        if admin:
+            self.relation_name = "db-admin"
+        else:
+            self.relation_name = "db"
+
+        super().__init__(charm, self.relation_name)
+
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_departed, self._on_relation_departed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.admin = admin
+        self.charm = charm
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the legacy db/db-admin relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        logger.warning(f"DEPRECATION WARNING - `{event.relation.name}` is a legacy interface")
+
+        unit_relation_databag = event.relation.data[self.charm.unit]
+        application_relation_databag = event.relation.data[self.charm.app]
+
+        # Do not allow apps requesting extensions to be installed.
+        if "extensions" in unit_relation_databag or "extensions" in application_relation_databag:
```

Comment:
I found some code to install extensions in the old pgbouncer charm. Do you want me to reject extensions in a similar way in the new charm, or pass them on into the postgres relation?

Reply:
Yes. For now we should also reject extensions on PgBouncer in a similar way.